### DISCUSSION TITLE
Using Annotated on Lists leads to crash

### DISCUSSION THREAD

**Discussion by secretrobotron** [fastapi/typer#861] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorials in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import typer
from typing import List
from typing_extensions import Annotated

app = typer.Typer()

@app.command()
def test1(test_list: List[int] = typer.Option(default=[1, 2, 3, 4])) -> None:
    for i in test_list:
        print(i)

@app.command()
def test2(test_list: Annotated[List[int], typer.Option(default=[1, 2, 3, 4])]) -> None:
     for i in test_list:
         print(i)

app()
```

### Description

While trying to work around [this problem](https://github.com/tiangolo/typer/issues/106), I ran into a new one. I have (I think) used `Annotated` correctly, but when I run the example code, I get this error:

```
AttributeError: 'list' object has no attribute 'isidentifier'
```

If I comment out `test2`, it works fine.

The traceback suggests the problem might be within Click, but I'm unsure:
```
 site-packages/click/core.py:2651 in _parse_decls  
```

### Operating System

macOS

### Typer Version

0.9.4

### Python Version

3.11.7

### Additional Context

Similar issues:

- https://github.com/tiangolo/typer/discussions/834
- https://github.com/tiangolo/typer/discussions/713

**Comment by svlandeg** [comment, marked as answer]
Hi!

You have to declare the default value slightly differently with `Annotated`. This should work as expected:

```
import typer
from typing import List
from typing_extensions import Annotated

app = typer.Typer()


@app.command()
def test2(test_list: Annotated[List[int], typer.Option()] = [1, 2, 3, 4]) -> None:
    for i in test_list:
        print(i)


if __name__ == "__main__":
    typer.run(test2)
```

**Reply by ShihaoXu** [reply]
Hi, the line
```
def test2(test_list: Annotated[List[int], typer.Option()] = [1, 2, 3, 4]) -> None:
```
isn't a perfect solution for the question.
I have two follow up questions:
1. `[1, 2, 3, 4]` is only initialized once and the same list would be used whenever `test2` is called. How can we defined a `default_factory` for the annotated function parameter?
2. It seems that `var: Annotated[X, typer.Option(default=Y)]` behaves differently than `var: Annotated[X, typer.Option()] = Y` in some situations. Is it a desired behavior?

**Reply by YuriiMotov** [reply]
> How can we defined a default_factory for the annotated function parameter?

You can specify default factory as:
```py
def test2(test_list: Annotated[List[int], typer.Option(default_factory=lambda: [1, 2, 3, 4])]) -> None:
```

> It seems that var: Annotated[X, typer.Option(default=Y)] behaves differently than var: Annotated[X, typer.Option()] = Y in some situations.

Could you please clarify what situations you are talking about?

**Comment by thht** [comment]
I think I have a similar situation here. The following code runs fine when using typer 0.16 and pinning click to <8.2 but fails with `TypeError: Secondary flag is not valid for non-boolean flag.` on click >=8.2. I hunted it down a bit: Apparently, typer just forwards the `default_factory` as the default value so the type for click ends up being `STRING`, which then leads to the confusion.

```python
def get_hashed_default() -> bool:
    config = Config.from_cmdargs()  # This needs typer initialized to work. So we need the default to be a callable.
    return config.hash_subject_ids

@app.command
def foo(hash_subject_ids: Annotated[bool, typer.Option(
        help='Whether to hash the subject IDs. Overrides the `hash_subject_ids` setting.',
        default_factory=get_hashed_default,
        show_default=False,
        ),
    ]
)
```

**Reply by svlandeg** [reply]
@thht : that sounds like a different issue, thanks for the report! We'll look into it.

**Reply by svlandeg** [reply]
@thht : I can't immediately replicate, can you give a full code file that works, the command you ran and the exact output you're getting?

**Reply by saandaar** [reply]
### Install dependencies:

```bash
pip install typer==0.16.1 click==8.2.2
```

### Example script:

```python
import typer
from typing_extensions import Annotated

app = typer.Typer()

def get_default_bool() -> bool:
    """A simple default factory that returns a boolean."""
    return True

@app.command()
def test_command(
    flag: Annotated[bool, typer.Option(
        "--flag/--no-flag",
        help="A boolean flag with default_factory",
        default_factory=get_default_bool,
        show_default=False,
    )]
):
    """Test command that reproduces the Click/Typer compatibility issue."""
    print(f"Flag value: {flag}")

if __name__ == "__main__":
    app()
```

### Run script:

```bash
python test.py test-command --help

# Output
# ...
# TypeError: Secondary flag is not valid for non-boolean flag.
```

**Reply by svlandeg** [reply]
@saandaar: thanks! This code snippet runs fine for me with both Click 8.1.8 and 8.2.1 and indeed crashes with 8.2.2. You should update your click version, as 8.2.2 has actually been [yanked](https://pypi.org/project/click/8.2.2/) due to "Unintended change in behavior of boolean options and None".

**Reply by thht** [reply]
Ok, thanks for looking into it!
